Services/UICore/classes/class.ilCtrlStructure.php

```diff
@@ -1,106 +1,318 @@
 <?php
-/* Copyright (c) 2020 Richard Klees, Extended GPL, see docs/LICENSE */
 
 /**
- * The CtrlStructure knows how GUIs call each other and where the code of the
- * guis is located.
+ * Class ilCtrlStructure holds the currently read control
+ * structure.
+ *
+ * @author Thibeau Fuhrer <thf@studer-raimann.ch>
  */
-class ilCtrlStructure
+final class ilCtrlStructure implements ilCtrlStructureInterface
```

Comment:
I actually dont know if this class is the new 'ilCtrl'. In the past i needed to extend this. I know the implications of extending such base classes a lot. I promise i dont write bug reports for extending this :) Does it really need to be a final class?

Reply:
May I ask why you needed to extend `ilCtrlStructure`? IMO you don't have to manipulate the call-structure, as it's read properly if all the necessary `@ilCtrl_Calls` or `@ilCtrl_isCalledBy` statements were made.

Reply:
I did that to make a fancy routing plugin. See: https://training.mtm.org/kursangebot/firmenspezifisch-11.html

So youre able to overwrite all ilias routes to make them static. This requires a lot of chained hacks to make it even work. Its nothing you want in the Core =)

I talked with @mjansenDatabay about this thuesday and he also showed me a plugin where he used a hack to manipulate some Tabs in a Ilias Object. 

If this class would be final, what would be the possibility to make such things happen? I actually would change the final on runtime or even copy the whole class into a plugin.

Reply:
What I did (in a plugin) was:

- Create an anonymous subclass of `ilCtrl` (which is possible because it is currently not final), which overrides all `get(LinkTaget|FormAction)(byClass)?` methods
- These methods don't call the parent method, but put all arguments in a map by using a hash as key, where this hash string is also returned to the caller
- Inject this into the `DIC`
- Call a method of `ilObj*GUI` to populate `ilTabsGUI` with all tabs of a repository plugin from an `UserInterfaceHook` plugin context
- Because I exchanged the `ilCtrl` instance, retrieving the tabs did not result in an error
- I could display them in the user interface of my plugin after I processed the stacked `ilCtrl` information, without being in a real (e.g.) course context

Reason for this: Repository objects have no mechanism/interface to expose theirs user interface tabs

Reply:
BTW: I removed the final keyword in the latest version

Reply:
Thank u

Reply:
There should be a discussion in the jourfix/technical board about the general usage of final. But i am not the person who decides that.

Reply:
Hi @PurHur,

this sounds like an interesting topic. May I ping you for a presentation weighing pros and cons (or something similar) for the next Dev-Conf in ~March?

Best regards!